CONTRIBUTING.txt

```diff
@@ -0,0 +1,98 @@
+    3. Problems are discrete, well-defined limitations or bugs.
+    4. FreeCAD uses the platform's issue-tracking system to track problems
+    5. Patches are sets of code changes that resolve a single problem
+    6. FreeCAD uses the Pull Request workflow for evaluating and accepting patches
+
+
+3. Roles
+    1. "User": An member of the wider FreeCAD community who uses the software.
+    2. "Contributor":  A person who submits a patch that resolves a previously identified problem. Contributors do not have commit access to the repository unless they are also maintainers. Everyone, without distinction or discrimination, SHALL have an equal right to become a Contributor
+    3. "Maintainer": A person who merges patches. Maintainers may or may not be developers. Their role is to enforce the process. Maintainers have commit access to the repository.
+    4. "Administrator": Administrators have additional authority to maintain the list of designated maintainers.
+
+4. Licensing, Ownership, and Credit
+    1. FreeCAD uses the Lesser General Public License, version 2, or superior (LGPL2+) for all source code.  Additional details can be found in the LICENSE file.
+    2. All contributions to FreeCAD SHALL use the same license as the project.
+    3. All contributions are owned by their authors unless assigned to another.
+    4. FreeCAD does not have a mandatory copyright assignment policy.
+    5. A Contributor who wishes to be identified in the Credits section of the application "About" dialog is responsible for identifying themselves appropriately by notifying an administrator or maintainer.
+
+5. Contribution Requirements
+
+    1. Patches are submitted in the form of Pull Requests (PR)
+    2. Maintainers and Contributors MUST have a platform account and SHOULD use their real names or a well-known alias. If the platform alias differs from the account alias on the FreeCAD Forum, effort SHOULD be taken to avoid confusion.
+    3. A PR SHOULD be a minimal and accurate answer to exactly one identified and agreed-on problem.
+    4. A PR SHOULD refrain from adding additional dependencies to the FreeCAD project unless no other option is available.
+    5. Code submissions MUST adhere to the code style guidelines of the project if these are defined.
+    6. If a PR contains multiple commits, each commit MUST compile cleanly and add value to the history of the project. Checkpoint commits should be squashed.
+    7. A PR SHALL NOT include non-trivial code from other projects unless the Contributor is the original author of that code.
+    8. A PR MUST compile cleanly and pass project self-tests on all target platforms.
+    9. Each commit message in a PR MUST succinctly explain what the commit achieves.
+    10. The PR Description MUST consist of a single short (less than 50 characters) line stating the problem (“Problem: …") being solved, followed by a blank line and then the proposed solution (“Solution: …").
```

Comment:
This seems to change our existing format to drop the [Workbench] prefix -- is that intended?

Reply:
I am in favor of letting the "Solution" text be as long and complex as necessary to describe the solution.  That should allow for multiple commit descriptions.

Reply:
Why the restriction to less than 50 characters? Could we use 72 characters here? This is a widely used standard for line wrapping and fits into most terminals. Especially if we would decide to keep some sort of [Prefixes], the remaining space for a description is further reduced and a meaningful description, which says something about the PR, becomes difficult.

Reply:
I don't see a reason why "Problem: " should be part of the PR _title_, but it being part of the description is totally fine. @chennes, am I missing something? How does this drop the [Workbench] prefix?

@sliptonic, the title prefixes should exist somewhere in the title/description of the issue because how else can the tags be set? I do believe I have seen issues where it wasn't clear what workbench(es) it it involves.

Reply:
@waebbl That's a git thing : from the git commit docs:

> Though not required, it’s a good idea to begin the commit message with a single short (less than 50 character) line summarizing the change, followed by a blank line and then a more thorough description. 

Because this is in the official docs, a lot of git-oriented tools rely on it for formatting (despite it not being "required").

Reply:
PRs are auto tagged based on changed files and you are asked to select a category when creating an issue, but I think this doesn't get auto tagged, it may be possible to make a gh action to auto tag those too.

Reply:
putting the problem _and_ the solution in the title seems excessive to me, and may be hard if we have only 50 char, IMO the solution should be rather explained in the commit messages and/or the PR description.

for example:
PR title: `Fix #XXXX (if an issue exists): short problem description
PR description:

> the issue was caused by this and that and can be fixed by this and that [...]

commit message:

> Fix #XXXX (if an issue exists): short problem description
> 
> -refactor something
> -fix some wrong conditionals
> -etc

Reply:
> IMO the solution should be rather explained in the commit messages and/or the PR description.

I like this.  Can you edit the document with proposed text that I can merge?  You can do this right in the web page.

Reply:
I think point 9 covers that actually and I see now that point 10 is talking about the PR **description** not title, and doesn't impose a character limit on the solution explanation. I would drop the "Problem: " and "Solution: " suffix though and 50 char is excessively short, this PR's title is longer than that and we have longer but still sensible titles: https://github.com/FreeCAD/FreeCAD/pull/7601
I'm not sure if we really need rules for PR descriptions, the rule above regarding commit messages is the important part IMO, and in any case it is there where we should impose a character limit on the first line.

Reply:
Do we all agree with the meaning of the terms `title`, `description` and `solution`? Those aren't the standard terms usually used for it. AFAICS the official git documentation states that the first line of the commit message is called title

> The text up to the first blank line in a commit message is treated as the commit title

We use the description for this. The git documentation uses description for the body of the commit message, as stated in message https://github.com/FreeCAD/FreeCAD/pull/7638/files#r1004540047

What's the situation if a PR consists of more than one commit? For one commit GitHub takes title and body from the commit message. If there are more than one commit, GitHub suggests a title, it somehow composes from terms in the commit titles and leaves the body empty.

How about this suggestion?
```
9. Each commit message in a PR MUST succinctly explain what the commit achieves. The commit message SHALL follow the suggestions in the `git commit --help` documentation, section DISCUSSION.
10. The PR message MUST consist of a single short line, the PR Title, summarizing the problem being solved, followed by a blank line and then the proposed solution in the Body. If a PR consits of more than one commit, the PR Title MUST succinctly explain what the PR achieves. The Body MAY be as detailed as needed.
```

Reply:
ok.  I'll change it.